### DISCUSSION TITLE
P/Invoke and a single UTF-8 character

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/runtime#98978] [Q&A] [answered]
All of the examples I have seen on the internet are about strings. But the C function I need to work with only takes a single character as input. The C side expects UTF-8. But .NET `char` is UTF-16.

**Comment by huoyaoyuan** [comment]
The concept of "single character" can be defined as multiple ways:
- A C `char`, which is capable to store ASCII character.
- A Unicode rune, which can be represented by 1 or 2 UTF-16 units (surrogate pair), or up to 3 bytes of UTF-8.
- A complex character that is composed by multiple runes.

**Comment by colejohnson66** [comment]
Are you asking how to P/Invoke to a method that takes a `char` parameter? You'd marshal as a `byte`.

For example, if your C method is `void Method(char c)`, you'd use:

```cs
[LibraryImport("<path>")]
public static partial void Method(byte c);
```

**Reply by danmoseley** [reply]
They mention UTF-8..

**Reply by colejohnson66** [reply]
Yes. C *strings* are UTF-8 (if `LC_ALL=xx_xx.UTF-8` or similar), but C *characters* are [just bytes](https://en.cppreference.com/w/cpp/language/types). The question says that "[...] the C function I need to work with only takes a single character as input." That, to me, sounds like it wants a single byte of data.

**Reply by tannergooding** [reply]
Just note that `char` (in C/C++) is implementation defined as to both its size and sign. Most implementations treat it as `8-bits`, and that is a reasonable expectation for anything .NET runs against, but not all platforms have done that historically.

For the sign, its commonly the case that implementations treat the sign the differently (including based on target architecture). For some platforms, the correct mapping is therefore `byte` and for others `sbyte`. This can actually impact how comparisons, upcasting, and shifting behave and can lead to subtle bugs/differences.

You can see some of these differences on [godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGe1wAyeAyYAHI%2BAEaYxCBmpAAOqAqETgwe3r56icmOAkEh4SxRMXF2mA6pQgRMxATpPn5ctpj2uQxVNQT5YZHRsbbVtfWZTQqDXcE9RX1mAJS2qF7EyOwc5gDMwcjeWADUJutuyGP4ggfYJhoAgpdXEaieuwCyXBDICDW7yLP7AOwWu2ImAISwYX32h12GgOAJMvwAIrdbvdHi8IMlgCF0F8PsQvj84QCgSDiGDkBC3FCYX9EddkQ9aM9Xl4GBisTjPt8/kTgaDwQdKdD1rCEbcOPNaJwAKy8PwcLSkVCcNzWay7BSLZaYfZmdY8UgETTi%2BYAa1ikgAdOtJJIAJx6sy/KS/X4ANl%2B%2Bk4kllRsVnF4ChAGgNRvmcFgMEQKFQLHidGi5EoaFj8Zi20MwAA%2BgRiCyTXw6ARooGIBFfRFgjUAJ6cfUV5jEKsAeQi2nKhu4vGTbEETYYtBr8t4WAiXmAbjEtEDndIWBYGfEQ9neCBFQAbphpwrMKpyl4i7XeMEi5Kl7Q8BFiNWPFhDwbiHgWIf5lQDMAFAA1PCYADuTfijB3vwggiGI7BSDIgiKCo6hLroTQGEYKCqpY%2BgXoGkDzKg8RtNOAC0TbrLwqAbsQD5YBhEDzGUFTOBArjDI0pCBJMhTFFkSQpAIjEcTkqTdGxfSjC07aVOMPHCa0YmdAJvQxKM4meA0ehjDJrFyRI1GaismmehwMqkHKCpKhwuyqAAHK6eGupIXyIcAuwQDmeY/BAuCECQOp6rMvAdlosymiAkhmFaZhcL86xSustpSlw5m2hoXCunp3qkE%2BUrBkZxH%2BrYQYhkOYaRhASDJnG9BkBQbwxmVfRrsg8TxJma5cLamYGEWYyZqoNkFrQRbECWZZLvW1Z3iNjYtm2Dh3t2jAEH2A6%2BiOY4TrQU53nOC6rAq%2BCro4G5brwO57geM7Hi0vrnpe14YKs%2Bo5o%2Bz58G%2Bn7fn%2BAFyvqwHCKI4gQd90FqL6ujSPZyGWNYaERJRWE4ak%2BFNmYxGkeRm7wNRIm0X49EMO4SkjMxuOydM8kJJxbQSWTfF5OpJMqZjbQdEM%2BNMTRjPjMT7EKZ0lOqbUnNCVpSw6VwErSj6S4mWZNm7CwCh1bszW2ha7WbgQjnuUQeIbKLvmhmLHCpelmW%2BiZAZ5X5xqkGawWheFkXRbF8WJclp5EYZps5ZbhXwOGUalamiZVSm5UgDULBruZeHpkYeHOQw%2BY0H1xaUENCrjYOdaVhNrbtjNMY9vN/aDjtmCjuOk7Tvqm1GIuO0rqJB2%2BsdyD7ndR6CBdZ4XlejY3u396PZ2L4vV%2Bv7/oBM7faBf3SADShA3B/Rg6YEOoVdMOKnDAgI%2B7JHRKjm9s6kLi45TLEFBpTTZFxaQs7xt8C6Tx8CEzdT35JomvxztNcwMPMf3/vzX%2BgsFjC3AgbAyWU/SmXDpHPCTA7IZkcvHE0rlNaeR1j5fK/lArrHWFafBRDiHEJSrwY2HtJY5UDMGb2pA/bFWjCHBMlUA6hzqg1JqLU2pMA6gQTM4cepJ36oNcs2dM68AzpNPOM5Zq9mLktMuK1K4bUwPOWu21hwN3XGjJcLc253nOqeBUV1e5Vn7neB6T5h7PSYO%2BMe71J5fVkDPcCc9ZCA1ggqXQ6x9AZnBlYde6F0Zb1wpwAiSNFQozwBREJL9sYMUARfKYf8b4U0AWk/iIDn4M2kszDIrNcnfzUpfOm3N8nKXKRMUpXMhZal0qeKBntYHEBYDLOWCslYq14WrDW%2BAtZeV1jgq2CBMBMCwDEKiZC0ogAypQ4y1CLb62tiAfBhCSEbN8W7CWCyOB6wKgbSJ0CzbDICqQUiyRnCSCAA%3D):
* Windows (MSVC) treats it as `sbyte` (signed) for `x86`, `x64`, and `Arm64`. 
* Unix (Clang/GCC) treats it as `sbyte` (signed) for `x86` and `x64` while it treats it as `byte` (unsigned) for Arm64.

**Reply by MichalPetryka** [reply]
You can also change it's signedness to with `-funsigned-char`.

**Reply by tannergooding** [reply]
Changing the signedness of char is notably dangerous and may subtly break code or other expectations. It is perhaps surprisingly common in various x86/x64 code to do something like `if (c >= 0)` to essentially do an `IsAscii` check.

In general its recommended to not do it where possible.

**Reply by ghost** [reply]
> You can also change it's signedness to with `-funsigned-char`.

The C runtime is still built with signed char. This will not work. Need to recompile the C runtime with this option, too. Not a solution.

**Comment by ghost** [comment]
It's an UTF-8 character. For example, the Euro sign. Not a simple ASCII character.

**Comment by elgonzo** [comment, marked as answer]
The answer to your question/problem essentially entirely depends on the C function parameter data type that receives the 1..4 UTF-8 code units for a char/rune...

**Reply by ghost** [reply]
I think I will have to write a wrapper function in C and P/Invoke that function instead.

**Reply by tannergooding** [reply]
Why not just have the managed side take `byte` or `sbyte` as appropriate, so that it maps down to `1-byte` on the native side.

For everything .NET works today, this works fine. You only really need to be careful if there is an upcast somewhere that would cause it to zero-extend when it should have sign-extended or vice versa.
